plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
+      //  with the rest of the cleanup: we'll be shutting down soon anyway due to bad_alloc
+      if(e)
+         std::rethrow_exception(e);
+      //coros always return on the session's strand
+      if(--coros_running == 0)
+         on_done(this);
    }
 
-   void process(state_history::get_blocks_request_v0& req) {
-      fc_dlog(plugin.get_logger(), "received get_blocks_request_v0 = ${req}", ("req", req));
-
-      auto self = this->shared_from_this();
-      auto entry_ptr = std::make_unique<blocks_request_send_queue_entry<session>>(self, std::move(req));
-      session_mgr.add_send_queue(std::move(self), std::move(entry_ptr));
+   template<typename F>
+   void drop_exceptions(F&& f) {
+      try{ f(); } catch(...) {}
    }
 
-   void process(state_history::get_blocks_request_v1& req) {
-      fc_dlog(plugin.get_logger(), "received get_blocks_request_v1 = ${req}", ("req", req));
-
-      auto self = this->shared_from_this();
-      auto entry_ptr = std::make_unique<blocks_request_send_queue_entry<session>>(self, std::move(req));
-      session_mgr.add_send_queue(std::move(self), std::move(entry_ptr));
-   }
+   template<typename F>
+   boost::asio::awaitable<void> readwrite_coro_exception_wrapper(F&& f) {
+      coros_running++;
 
-   void process(state_history::get_blocks_ack_request_v0& req) {
-      fc_dlog(plugin.get_logger(), "received get_blocks_ack_request_v0 = ${req}", ("req", req));
-      if (!current_request) {
-         fc_dlog(plugin.get_logger(), " no current get_blocks_request_v0, discarding the get_blocks_ack_request_v0");
-         return;
+      try {
+         co_await f();
       }
-
-      auto self = this->shared_from_this();
-      auto entry_ptr = std::make_unique<blocks_ack_request_send_queue_entry<session>>(self, std::move(req));
-      session_mgr.add_send_queue(std::move(self), std::move(entry_ptr));
-   }
-
-   state_history::get_status_result_v0 get_status_result() {
-      fc_dlog(plugin.get_logger(), "replying get_status_request_v0");
-      state_history::get_status_result_v0 result;
-      result.head              = plugin.get_block_head();
-      result.last_irreversible = plugin.get_last_irreversible();
-      result.chain_id          = plugin.get_chain_id();
-      auto&& trace_log         = plugin.get_trace_log();
-      if (trace_log) {
-         auto r = trace_log->block_range();
-         result.trace_begin_block = r.first;
-         result.trace_end_block   = r.second;
+      catch(std::bad_alloc&) {
+         throw;
       }
-      auto&& chain_state_log = plugin.get_chain_state_log();
-      if (chain_state_log) {
-         auto r = chain_state_log->block_range();
-         result.chain_state_begin_block = r.first;
-         result.chain_state_end_block   = r.second;
+      catch(fc::exception& e) {
+         if(has_logged_exception.test_and_set() == false)
+            fc_ilog(logger, "state history connection from ${a} failed: ${w}", ("a", remote_endpoint_string)("w", e.top_message()));
```

Comment:
I'm uncertain what "exceptions below" you're referring to here. An example message that could occur here would be
```
state history connection from 127.0.0.1:57580 failed: Provided index out of range for variant
```
(I sent some trash bytes as a request instead of a value `request` variant.

We generally don't display the multi-line verbose fc error stacks in a log message I don't think? (if that's what you're referring to)

Reply:
I saw each exception starting with "state history connection from..." and was thinking it would be hard to tell from which exception. But you have detailed information after "state history connection from...". It should be easy to tell which exception happened.